src/backend/x11/window.rs

```diff
@@ -0,0 +1,259 @@
+
+        // Enable WM_DELETE_WINDOW so our client is not disconnected upon our toplevel window being destroyed.
+        xcb.change_property32(
+            PropMode::REPLACE,
+            window.inner,
+            atoms.WM_PROTOCOLS,
+            AtomEnum::ATOM,
+            &[atoms.WM_DELETE_WINDOW],
+        )?;
+
+        // Block until window creation is complete.
+        cookie.check()?;
+        window.set_title(properties.title)?;
+
+        // Finally map the window
+        xcb.map_window(window.inner)?;
+
+        // Flush requests to server so window is displayed.
+        xcb.flush()?;
+
+        Ok(window)
+    }
+
+    pub fn map(&self) -> Result<(), X11Error> {
+        self.connection.xcb_connection().map_window(self.inner)?;
+        Ok(())
+    }
+
+    pub fn unmap(&self) -> Result<(), X11Error> {
+        let xcb = self.connection.xcb_connection();
+
+        // ICCCM - Changing Window State
+        //
+        // Normal -> Withdrawn - The client should unmap the window and follow it with a synthetic
+        // UnmapNotify event as described later in this section.
+        xcb.unmap_window(self.inner)?;
+
+        // Send a synthetic UnmapNotify event to make the ICCCM happy
+        xcb.send_event(
+            false,
+            self.inner,
+            EventMask::STRUCTURE_NOTIFY | EventMask::SUBSTRUCTURE_NOTIFY,
+            UnmapNotifyEvent {
+                response_type: x11rb::protocol::xproto::UNMAP_NOTIFY_EVENT,
+                sequence: 0, // Ignored by X server
+                event: self.root,
+                window: self.inner,
+                from_configure: false,
+            },
+        )?;
+
+        Ok(())
+    }
+
+    pub fn size(&self) -> Size<u16, Logical> {
+        *self.size.lock().unwrap()
+    }
+
+    pub fn set_title(&self, title: &str) -> Result<(), X11Error> {
+        let xcb = self.connection.xcb_connection();
+
+        // _NET_WM_NAME should be preferred by window managers, but set both in case.
+        xcb.change_property8(
+            PropMode::REPLACE,
+            self.inner,
+            AtomEnum::WM_NAME,
+            AtomEnum::STRING,
+            title.as_bytes(),
+        )?;
+
+        xcb.change_property8(
+            PropMode::REPLACE,
+            self.inner,
+            self.atoms._NET_WM_NAME,
+            self.atoms.UTF8_STRING,
+            title.as_bytes(),
+        )?;
+
+        // Set WM_CLASS
+        let mut raw = Vec::new();
+        raw.extend_from_slice(title.as_bytes());
+        raw.extend_from_slice(b"\n");
+        raw.extend_from_slice(title.as_bytes());
+        raw.push(b'\n');
+
+        let _ = xcb.change_property8(
+            PropMode::REPLACE,
+            self.inner,
+            self.atoms.WM_CLASS,
+            AtomEnum::STRING,
+            &raw[..],
+        )?;
+
+        Ok(())
+    }
+}
+
+impl Drop for WindowInner {
+    fn drop(&mut self) {
+        let _ = self.connection.xcb_connection().destroy_window(self.inner);
+    }
```

Comment:
Yay, another case where the new `WindowWrapper` might be slightly useful (I already mentioned this above for `PixmapWrapper`).